inc/sitemaps/class-sitemaps.php

```diff
@@ -0,0 +1,429 @@
+		if ( is_scalar( $current_page ) && intval( $current_page ) > 0 ) {
+			$this->current_page = intval( $current_page );
+		}
+	}
+
+	/**
+	 * Set the sitemap content to display after you have generated it.
+	 *
+	 * @param string $sitemap The generated sitemap to output.
+	 */
+	public function set_sitemap( $sitemap ) {
+		$this->sitemap = $sitemap;
+	}
+
+	/**
+	 * Set a custom stylesheet for this sitemap. Set to empty to just remove the default stylesheet.
+	 *
+	 * @deprecated
+	 *
+	 * @param string $stylesheet Full xml-stylesheet declaration.
+	 */
+	public function set_stylesheet( $stylesheet ) {
+		$this->renderer->set_stylesheet( $stylesheet );
+	}
+
+	/**
+	 * Set as true to make the request 404. Used stop the display of empty sitemaps or invalid requests.
+	 *
+	 * @param bool $bool Is this a bad request. True or false.
+	 */
+	public function set_bad_sitemap( $bool ) {
+		$this->bad_sitemap = (bool) $bool;
+	}
+
+	/**
+	 * Prevent stupid plugins from running shutdown scripts when we're obviously not outputting HTML.
+	 *
+	 * @since 1.4.16
+	 */
+	public function sitemap_close() {
+		remove_all_actions( 'wp_footer' );
+		die();
+	}
+
+	/**
+	 * Hijack requests for potential sitemaps and XSL files.
+	 *
+	 * @param \WP_Query $query Main query instance.
+	 */
+	public function redirect( $query ) {
+
+		if ( ! $query->is_main_query() ) {
+			return;
+		}
+
+		$xsl = get_query_var( 'xsl' );
+
+		if ( ! empty( $xsl ) ) {
+			$this->xsl_output( $xsl );
+			$this->sitemap_close();
+
+			return;
+		}
+
+		$type = get_query_var( 'sitemap' );
+
+		if ( empty( $type ) ) {
+			return;
+		}
+
+		$this->set_n( get_query_var( 'sitemap_n' ) );
+
+		/**
+		 * Filter: 'wpseo_enable_xml_sitemap_transient_caching' - Allow disabling the transient cache
+		 *
+		 * @api bool $unsigned Enable cache or not, defaults to true
+		 */
+		$caching = apply_filters( 'wpseo_enable_xml_sitemap_transient_caching', true );
+
+		if ( $caching ) {
+			do_action( 'wpseo_sitemap_stylesheet_cache_' . $type, $this );
+			$this->sitemap   = get_transient( 'wpseo_sitemap_cache_' . $type . '_' . $this->current_page );
+			$this->transient = ! empty( $this->sitemap );
+		}
+
+		if ( empty( $this->sitemap ) ) {
+			$this->build_sitemap( $type );
+		}
+
+		if ( $this->bad_sitemap ) {
+			$query->set_404();
+			status_header( 404 );
+
+			return;
+		}
+
+		if ( $caching && ! $this->transient ) {
+			set_transient( 'wpseo_sitemap_cache_' . $type . '_' . $this->current_page, $this->sitemap, DAY_IN_SECONDS );
+		}
+
+		$this->output();
+		$this->sitemap_close();
+	}
+
+	/**
+	 * Attempts to build the requested sitemap.
+	 *
+	 * Sets $bad_sitemap if this isn't for the root sitemap, a post type or taxonomy.
+	 *
+	 * @param string $type The requested sitemap's identifier.
+	 */
+	public function build_sitemap( $type ) {
+
+		$type = apply_filters( 'wpseo_build_sitemap_post_type', $type );
+
+		if ( $type === '1' ) {
+			$this->build_root_map();
```

Comment:
I feel `build_root_map` is of the same order as `build_sitemap`. I'd move the conditional logic in which this function is called one level up. So then it would end up in `WPSEO_Sitemaps::redirect` (which then still needs to be broken down further). The `wpseo_build_sitemap_post_type` filter could also be called straight on the `$type` assignment in there. Doing this will give `WPSEO_Sitemaps::build_sitemap` a more clear-cut purpose.